Ported the vrf verification feature

adon-at-work opened the pull request:
This is the vrf verification implementation ported from https://github.com/yahoo/coname/blob/master/vrf/vrf.go 

It lives as an utility function, and is tested with inputs and outputs captured from the original golang implementation.

Please review, thanks.

Commit:
added sha3 SHAKE256 lib

- imported js-sha3 v0.3.1 from https://github.com/emn178/js-sha3
- extended js-sha3 to support SHAKE256 (output correct up to first
1088bits)
- allowed output byte array of 32 bytes

Commit:
added vrf verify implementation

- translated the golang implementation from
https://github.com/yahoo/coname/blob/master/vrf/vrf.go#L148

diracdeltas commented on src/javascript/crypto/e2e/vrf/sha3shake256.js:
```diff
@@ -0,0 +1,353 @@
+ * Copyright 2015, emn178@gmail.com
+ * 
+ * Licensed under the MIT license:
+ * http://www.opensource.org/licenses/MIT
```
Should this go under the `@license` header? I'm not sure what is the recommended way to include the MIT license in an Apache2 project. @gyehuda

adon-at-work replied:
gd point. i was curious what to do too. I copied the code from there, dropped some, and added some new features. so, what's the license then? follow the MIT? or stick with the Apache as used in other files? or dual licenses? trust the expert @gyehuda can give some insights. thanks

gyehuda replied:
Sorry for the delay. There are a couple of ways to address this, and it's mostly driven by project convention. Some projects have a strict process where the headers are kept consistent and that all inclusions like this are disclosed in a NOTICES file, other projects are much less strict and have what you see above -- an inclusion in the code that is preceded with the relevant license and copyright text. Both are OK. At first this latter approach is simpler, eventually the code could get messy and you'll want to clean it up. 

What is most important though is that somewhere in the project you indicate that emn17 owns the copyright of some code that was licensed under the MIT license and is included in this project.

adon-at-work replied:
Thanks @gyehuda. 
@diracdeltas, can you suggest which convention/approach you prefer?

diracdeltas commented on src/javascript/crypto/e2e/vrf/extra25519.js:
```diff
@@ -0,0 +1,185 @@
+goog.require('e2e.ecc.DomainParam');
+goog.require('e2e.ecc.PrimeCurve');
+
+goog.require('e2e.vrf.sha3Shake256');
```
Nit: requires should be in alphabetical order. You can run `./do.sh lint` on the newly-added files to find linting errors.

adon-at-work replied:
thanks. will fix today when I get back to office. i dont have that lint tool in my personal laptop. :)

adon-at-work replied:
the changes are prepared at https://gist.github.com/adon-at-work/ca254eed709ecf322d47
will commit it to this PR at last, as to avoid collapsing all the comments due to "outdated diff".

diracdeltas replied:
Ok great. I will run tests and do a build after that pull request.

andres-erbsen commented on src/javascript/crypto/e2e/vrf/extra25519.js:
```diff
@@ -0,0 +1,185 @@
+
+	var P = e2e.vrf.ed25519.curve.pointFromByteArray(sBytes);
+
+	goog.asserts.assert(!P.isInfinity(),
```
Are assertions the right tool here? I am used to using assertions to detect programmer errors, but a completely good user running perfect code could still be sent an invalid point by a malicious server.

adon-at-work replied:
I sort of follow the practise used by the ecc library. [Example 1](https://github.com/yahoo/end-to-end/blob/master/src/javascript/crypto/e2e/ecc/point/curve25519.js#L71-L72) [Example 2](https://github.com/yahoo/end-to-end/blob/master/src/javascript/crypto/e2e/ecc/curve/curve25519.js#L85)

adon-at-work replied:
I just had another look at https://github.com/yahoo/coname/blob/master/vrf/vrf.go#L171 and https://github.com/yahoo/coname/blob/master/ed25519/edwards25519/edwards25519.go#L1095-L1127, even a check failed and returned `false`, [p is already filled](https://github.com/yahoo/coname/blob/master/ed25519/edwards25519/edwards25519.go#L1103) and your code is apparently ignoring what is returned. Is that expected?

andres-erbsen replied:
That is the code I wrote and intended to write. In retrospect it would have been nicer to actually zero the output when an error is detected (but no harm done since the returned value does not get used).

adon-at-work replied:
My observation is that: assume the following `pk` used in `FromBytesBaseGroup()` fail to satisfy `nP == IDENTITY`. 
The calculation will continue to produce the subsequent `ABytes`, and so forth...

``` go
P.FromBytesBaseGroup(&pk)
edwards25519.GeDoubleScalarMultVartime(&A, &c, &P, &t)
A.ToBytes(&ABytes)
```

So, indeed the check on `nP == IDENTITY` (and the `P.toBytes(P.fromBytes(pk)) == pk`) is actually optional. 
Therefore, do you think I should just remove all the checks below as to conform to your implementation?

In fact, something like follows makes more sense to me:

``` go
if !P.FromBytesBaseGroup(&pk) {
    return false
}
```

Thoughts?

adon-at-work replied:
note: I changed the implementation from using assertion to throw

but please confirm whether the checks should be enforced in a way that when they failed, we fail the whole vrf verification by returning false.

andres-erbsen replied:
The checks are not optional, and are actually very important. My Go implementation was wrong. https://github.com/yahoo/coname/pull/26 is a thing that should happen (and soon after this PR it should also get tests).

adon-at-work replied:
Got you. I merged over there. :+1:

andres-erbsen commented on src/javascript/crypto/e2e/vrf/extra25519.js:
```diff
@@ -0,0 +1,185 @@
+	goog.asserts.assert(P.isOnCurve(),
+        'P must satisfy the defining equation of the elliptic curve E');
+
+	// TODO: confirm https://github.com/google/end-to-end/issues/340
```
I have no idea what is going on there. Multiplication by order really should give the identity element. This check is important, though -- we should not launch with it simply omitted.

adon-at-work replied:
Right now, `P.multiply(e2e.vrf.ed25519.n).isInfinity()` evaluates to `false`, whereas `P.multiply(e2e.vrf.ed25519.n).isIdentity()` evaluates to `true`.

The underlying function calls can be found at here: [isInfinity(), isIdentity()](https://github.com/yahoo/end-to-end/blob/master/src/javascript/crypto/e2e/ecc/point/ed25519.js#L162-L174), and the points [INFINITY, IDENTITY](https://github.com/yahoo/end-to-end/blob/master/src/javascript/crypto/e2e/ecc/curve/ed25519.js#L103-L117).

I was curious but didn't have time to figure out yet why the infinity point is different than the identity point under the ed25519 implementation, while they're equal under the [curve25519 one](https://github.com/yahoo/end-to-end/blob/master/src/javascript/crypto/e2e/ecc/point/curve25519.js#L104-L118).

The check at the [golang implementation](https://github.com/yahoo/coname/blob/master/ed25519/edwards25519/edwards25519.go#L1117) is apparently checking against the identity point. So I wondered if `isIdentity()` is what we want to check at here (line 58). Same for line 48? Any ideas?

adon-at-work replied:
@andres-erbsen, I'm so glad we can have your reviews here. You're so awesome. thanks. :+1:

andres-erbsen replied:
Hmm. I think isIdentity is the way to go, and I am not sure what meaning `this.INFINITY = new e2e.ecc.point.Ed25519(this,  this.ZERO, this.ZERO, this.ZERO, this.ZERO)` has.

The reason we want this check is that we want all points used in the vrf computation to be in the prime-order subgroup of the non-prime-order elliptic curve. Multiplying by order of the subgroup should give the identity element (of both the subgroup and the curve) in that case. This reasoning does not involve the geometric ("at infinity") interpretation at all.

http://iml.univ-mrs.fr/ati/crypto_puces/2011/slides/Mardi/rolland.pdf and https://christianepeters.files.wordpress.com/2012/10/20080620-rennes.pdf say that projective(0,1,0) (the identity) is a point at infinity. https://hyperelliptic.org/EFD/oldefd/edwards.html and https://en.wikipedia.org/wiki/Edwards_curve#Projective_homogeneous_coordinates confirm that that is the neutral element (with no mention of infinity).

Traditional NIST curves only have one point at infinity and this is the identity. In edwards curves, we cannot assume that. Which still does not explain the value of `INFINITY` this library defines for ed25519, but oh well. I also find it interesting that the ed25519 code comments mention jacobian coordinates but reference the EFD entry for addition law in the (extended) projective coordinates.

adon-at-work replied:
agree. I switched to use `isIdentity()`.
And let's track the infinity point and `isInfinity()` that we're unsure at https://github.com/google/end-to-end/issues/340

andres-erbsen commented on src/javascript/crypto/e2e/vrf/extra25519.js:
```diff
@@ -0,0 +1,185 @@
+
+	// e is either 1 or -1
+	// x = ev - (1-e)A/2
+	var x = e.isEqual(e2e.vrf.ed25519.curve.ONE) ? v : A.substract(v);
```
Please document that this is not constant time and thus must not be used with secret inputs.

adon-at-work replied:
ok

andres-erbsen commented on src/javascript/crypto/e2e/vrf/sha3shake256.js:
```diff
@@ -0,0 +1,353 @@
+var blocks = [],
+    s = [];
+
+var keccak = function(message, bits, padding, outputByteLen, outputFormat) {
```
I am not sure what the constant-time status of this is. "yes" for non-strings and "no" for strings? Maybe just err on the side of safety and say it is not necessarily constant time.

adon-at-work replied:
we can just comment out the string implementation

andres-erbsen commented on src/javascript/crypto/e2e/vrf/vrf_test.html:
```diff
@@ -0,0 +1,53 @@
+aliceProof:   318004FD67F1A433152DA837126716E9F588F2EE71DAA07A81594867FADE0303EFEB5D62AD73A818DEA5BAE08A4CC9ED82C92F12BF183D50718BF6E9175EB10CC16A26AC42C016BCB10E906426B36046379D508B913E5EC3B5E0B72A404291A2
+*/
+
+var pk = [194, 191, 96, 139, 106, 249, 24, 253, 198, 131, 88, 169, 100, 231, 7, 211, 70, 171, 171, 207, 24, 30, 150, 114, 77, 124, 240, 123, 191, 14, 29, 111];
```
It would be great if the same test vector was in the Go library as well and they linked to each other in comments.

adon-at-work replied:
Do you expect something like the following: 
https://github.com/adon-at-work/coname/blob/debug/bytedump/vrf/vrf_test.go

I can submit a PR to add such a test case for the golang implementation

adon-at-work replied:
completed in https://github.com/yahoo/coname/pull/25

andres-erbsen commented on src/javascript/crypto/e2e/vrf/vrf_test.html:
```diff
@@ -0,0 +1,53 @@
+ * Tests that e2e.vrf.verify() works
+ */
+function testVRFVerify() {
+    assertTrue(e2e.vrf.verify(pk, alice, aliceVRF, aliceProof));
```
Please add a negative test.

adon-at-work replied:
agree. i actually tried to randomly alter a byte by hand, and it returned false. :+1:

adon-at-work replied:
addressed in https://github.com/yahoo/end-to-end/commit/5fb47531ee9ae5bbd44fc3ae745b20eed9443a8b#diff-565ba75ebd07abd46d0cb857452443b7

andres-erbsen commented on src/javascript/crypto/e2e/vrf/vrf.js:
```diff
@@ -0,0 +1,154 @@
+	// edwards25519.GeDoubleScalarMultVartime(&hmtP, &t, hm, &[32]byte{})
+	// edwards25519.GeDoubleScalarMultVartime(&iicP, &c, &ii, &[32]byte{})
+	var hm = hashToCurve(m);
+	var hmtP = e2e.vrf.extra25519.geDoubleScalarMultVartime(t, hm, e2e.BigNum.ZERO);
```
The Go code called `geDoubleScalarMultVartime` because it was faster than the single multiplication. This is probably not true here. Which does not necessarily mean that we want to change this code, but if ever performance becomes an issue, this is a place to look.

adon-at-work replied:
Agree. You have a better name for this function? or should I just remove this function call.

andres-erbsen replied:
I think `Vartime` means "not necessarily constant time", so the name is fine. I have no opinion on removing (by inlining and then simplifying).

andres-erbsen commented on src/javascript/crypto/e2e/vrf/vrf.js:
```diff
@@ -0,0 +1,154 @@
+}
+
+
+/**
```
This documentation does not seem to match the function it is next to.

adon-at-work replied:
gd catch. I noticed that and have planned to fix it as in:
https://gist.github.com/adon-at-work/ca254eed709ecf322d47#file-fixed-gjshint-errors-diff-L1101-L1115

andres-erbsen commented on src/javascript/crypto/e2e/vrf/extra25519_test.html:
```diff
@@ -0,0 +1,87 @@
+/**
+ * Tests that e2e.vrf.extra25519.fromBytesBaseGroup works, and no assertion error is thrown
+ */
+function testFromBytesBaseGroup() {
```
We actually want better tests for this; even just code coverage is better than nothing here. My code does not seem to have any either, sorry about that. I do not have time to implement this right now and I am not sure when I will. So I am just writing something here with no well-thought intent. Some things that are probably true, and if they are, would be useful to check:
- Any multiple of the base point should unmarshal correctly.
- The blacklisted points at http://cr.yp.to/ecdh.html#validate should return error (but the list there is in X25519 format, not ed25519)
- The identity point should return error.
- The points of order 2 and of order 4 from https://christianepeters.files.wordpress.com/2012/10/20080710-s3cm.pdf should return error.
- x_B + y_S should return error if B is the base point and S is a point of order 2 or order 4 and y is less than its order.

There is an additional complication that a normal elliptic curve library might not even care to support the "bad" points, so using them for testing might end up being awkward. In particular, multiplication y*S does clear the low bits of y in some libraries, which would not work here.

adon-at-work replied:
I prefer to leave this as a separate PR. OK?

andres-erbsen replied:
Yep, that was the intent.

adon-at-work replied:
tracking it at https://github.com/yahoo/end-to-end/issues/59

adon-at-work commented:
@andres-erbsen , once you finished reviewing this. I'll clear all the commented golang here. They're there to facilitate reviews only. OK?

Commit:
updated according to reviewer's comments

adon-at-work commented:
@andres-erbsen and @diracdeltas please check those things fixed.

Since many of those discussions are collapsed since the last commit, I list out the following discussions that are yet to be resolved for quick reference:
- https://github.com/yahoo/end-to-end/pull/58#discussion-diff-39409648
- https://github.com/yahoo/end-to-end/pull/58#discussion-diff-39459701
- https://github.com/yahoo/end-to-end/pull/58#discussion-diff-39517955

Commit:
match the latest golang implementation

Ref:
https://github.com/yahoo/coname/commit/671193d83ba82ba8f5aab3e4b411c8493
d089c12

adon-at-work commented:
After the last commit, the final issue to deal with in this PR: https://github.com/yahoo/end-to-end/pull/58#discussion-diff-39409648

Commit:
cleaned up sha3 impl. according to @andres-erbsen's comments

diracdeltas commented on src/javascript/crypto/e2e/vrf/vrf.js:
```diff
@@ -0,0 +1,170 @@
+  //  return false
+  // }
+  // hash.Reset()
+  hash = e2e.vrf.sha3Shake256(iiB.concat(m), 32);
```
hash is undeclared

adon-at-work replied:
gd catch

diracdeltas commented:
Currently, the VRF module isn't included in any of the extension files, so the compiler isn't actually checking it for correctness. When I include the vrf module in the extension (such as by prepending `goog.require('e2e.vrf.verify');` to `launcher.js`), Closure compiler throws a lot of errors.

diracdeltas commented on src/javascript/crypto/e2e/vrf/extra25519.js:
```diff
@@ -0,0 +1,178 @@
+goog.require('e2e.ecc.PrimeCurve');
+goog.require('e2e.error.InvalidArgumentsError');
+
+var ed25519 = e2e.ecc.DomainParam.fromCurve(e2e.ecc.PrimeCurve.ED_25519);
```
Declaring variables/functions like this will pollute the global namespace. Please make sure everything is contained in the module namespace. Ex: `e2e.vrf.extra255519.ed25519_ = e2e.ecc.DomainParam.fromCurve(e2e.ecc.PrimeCurve.ED_25519)`

The underscore at the end of a name indicates to the compiler that this variable is private (should not be accessed outside of the current module).

adon-at-work replied:
makes sense.

diracdeltas commented on src/javascript/crypto/e2e/vrf/vrf.js:
```diff
@@ -0,0 +1,170 @@
+goog.require('e2e.vrf.extra25519');
+goog.require('e2e.vrf.sha3Shake256');
+
+var PublicKeySize = 32,
```
Should these be in all-caps? (for constants)

adon-at-work replied:
I can change them as private variables for this namespace with all-caps.

diracdeltas commented on src/javascript/crypto/e2e/vrf/extra25519_test.html:
```diff
@@ -0,0 +1,56 @@
+function testFromBytesBaseGroup() {
+	var pk = [194, 191, 96, 139, 106, 249, 24, 253, 198, 131, 88, 169, 100, 231, 7, 211, 70, 171, 171, 207, 24, 30, 150, 114, 77, 124, 240, 123, 191, 14, 29, 111];
+
+	e2e.vrf.extra25519.fromBytesBaseGroup(pk);
```
this test should probably contain some sort of assert statement.

adon-at-work replied:
gd catch. assert statements were once inlined in `fromBytesBaseGroup()`, but later removed. will update it accordingly

Commit:
addressed @diracdeltas comments

adon-at-work commented:
> Currently, the VRF module isn't included in any of the extension files, so the compiler isn't actually checking it for correctness. When I include the vrf module in the extension (such as by prepending `goog.require('e2e.vrf.verify');` to `launcher.js`), Closure compiler throws a lot of errors.

the errors are cleared.

adon-at-work commented:
I'm done with the suggested changes.

The build was failed due to an upstream commit (happened roughly an hour ago):
https://github.com/google/closure-templates/commit/e47f247bdfacf3c5a34015ced356e4fd8a0444a2

Commit:
Fallback to use a buildable version of closure-templates compiler

diracdeltas commented on src/javascript/crypto/e2e/vrf/extra25519.js:
```diff
@@ -0,0 +1,192 @@
+ * @param {!e2e.ecc.Element} r the representative coordinate
+ * @return {!e2e.ecc.Element}
+ */
+function representativeToMontgomeryX(r) {
```
Please put this in the e2e.vrf.extra25519 namespace

diracdeltas commented on src/javascript/crypto/e2e/vrf/extra25519.js:
```diff
@@ -0,0 +1,192 @@
+ * @param {!e2e.ecc.Element} x the Curve25519 coordinate
+ * @return {!e2e.ecc.Element}
+ */
+function montgomeryXToEdwardsY(x) {
```
same as above

diracdeltas commented on src/javascript/crypto/e2e/vrf/sha3shake256.js:
```diff
@@ -0,0 +1,353 @@
+
+goog.require('goog.asserts');
+
+
```
all the variable declarations in the top-level namespace below should be in the e2e.vrf.sha3Shake256 namespace

diracdeltas commented on src/javascript/crypto/e2e/vrf/vrf.js:
```diff
@@ -0,0 +1,194 @@
+
+
+/**
+ * @private Public Key Size
```
nit: should be `PUBLIC_KEY_SIZE_`. Same for other constants below.

diracdeltas commented on src/javascript/crypto/e2e/vrf/vrf.js:
```diff
@@ -0,0 +1,194 @@
+ * @param {!e2e.ByteArray} m A byte array of length 32
+ * @return {!e2e.ecc.point.Ed25519}
+ */
+function hashToCurve(m) {
```
put in the e2e.vrf namespace

Commit:
dealt with namespacing

Commit:
upgrade to use the new SHA3 implementation

to leverage the faster new interfaces update() and digest()

adon-at-work commented:
@diracdeltas , all done. upgraded to use the latest js-sha3 too

diracdeltas commented:
Thanks, LGTM! @andres-erbsen any objections to merging?

andres-erbsen commented on src/javascript/crypto/e2e/vrf/extra25519.js:
```diff
@@ -0,0 +1,192 @@
+ * @param {!e2e.ByteArray} h A byte array of length 32
+ * @return {!e2e.ecc.point.Ed25519}
+ */
+e2e.vrf.extra25519.hashToEdwards = function(h) {
```
Please document that this is not constant time (because variable-length bignums aren't)

andres-erbsen commented on src/javascript/crypto/e2e/vrf/vrf.js:
```diff
@@ -0,0 +1,200 @@
+ * @param {!e2e.ByteArray} m A byte array of length 32
+ * @return {!e2e.ecc.point.Ed25519}
+ */
+e2e.vrf.hashToCurve = function(m) {
```
Please document that this is not constant time and thus should not be used with secret inputs. reason: hashToCurve isn't.

andres-erbsen commented:
No objections. I did add two one-line documentation nits about functions that are not constant-time-ish, though.
